### DISCUSSION TITLE
NextJS + Relay

### DISCUSSION THREAD

**Discussion by paulm17** [vercel/next.js#32502] [Ideas]
So upon watching this: https://www.youtube.com/watch?v=lhVGdErZuN4

Definitely excited about the opportunities with React 18, service side components and components having their own data providers.  Possibilities are endless so to speak.

Problem is, when I spent some time trying to get relay to work with NextJS it wasn't easy.  Here's the hurdles that I faced as a noob to the whole graphql stack.

1) Relay docs not friendly at all.
2) NextJS interoperability is a utility lib and way way too much boilerplate and issues with SSR abound.
3) Not a satisfying experience.
4) Can't compete in terms of ease of use and getting to work vs Prisma (ORM).

With this in mind, does Vercel have any future plans to make this much more simpler?

Thanks

**Comment by wvanrensselaer** [comment]
Looking into this a bit, particularly Relay + React 18 Streaming in Next.js to take advantage of SSR Suspense.

The `concurrentFeatures` flag opts into a sandboxed environment without access to Node APIs. [babel-plugin-relay](https://github.com/facebook/relay/tree/main/packages/babel-plugin-relay) uses [babel-plugin-macros](https://github.com/kentcdodds/babel-plugin-macros) which uses the Node [fs](https://nodejs.org/api/fs.html) API which doesn't work with incremental builds in dev mode in Next.js.

Couple solutions I can think of here:
- Port `babel-plugin-relay` and `babel-plugin-macros` to SWC plugins using [the new WASM based plugin system](https://github.com/swc-project/swc/issues/3167). I think Next.js might need to upgrade SWC to support this. Also needs custom SWC plugin support (https://github.com/vercel/next.js/discussions/30212). This will be nice anyway to allow apps to use SWC + Relay. Going to look into this a bit.
- Support Node APIs in the sandboxed environment.

There may also be other issues beyond this, but can't make it further until this is solved.

**Reply by hamlim** [reply]
The new [`runtime` config option](https://github.com/vercel/next.js/issues/31317) might allow us to get around this limitation!

**Reply by wvanrensselaer** [reply]
Tried out the `runtime` flag, works great for a production build, but still has issues with `babel-plugin-macros` using `fs` in dev mode. Can reproduce in this repo: https://github.com/wvanrensselaer/nextjs-relay-streaming

**Reply by roippi** [reply]
swc relay support has been merged to canary, I believe?  There's some back and forth in a few issues, seems like the extension is giving them some grief.  https://github.com/vercel/next.js/pull/33702

Edit: ah this was picked up in `v12.1.0` https://github.com/vercel/next.js/releases/tag/v12.1.0

**Reply by wvanrensselaer** [reply]
Yeah, that support appears to be partial to just `relay-compiler`. I pushed a branch to test it out: https://github.com/wvanrensselaer/nextjs-relay-streaming/tree/compiler-relay-use

1. It's not generating types, [unless I have it misconfigured](https://github.com/wvanrensselaer/nextjs-relay-streaming/blob/compiler-relay-use/app/next.config.js#L8).
2. Opting into the compiler option requires opting out of the Babel config. That support doesn't seem to include `babel-plugin-relay/macro` for query definitions. So still seeing the dev issue.

**Comment by wvanrensselaer** [comment]
Noting another limitation - Refetches data client side. Need to figure out how to hydrate Relay cache client side with streaming.

This is called out here: https://nextjs.org/docs/advanced-features/react-18/streaming#data-fetching

It's an upstream issue, Suspense does not yet support a client cache hydration mechanism: https://github.com/reactwg/react-18/discussions/22

> This API is not yet integrated with data fetching. The general Suspense mechanism should work, but we don't have a recommendation yet for how to transfer data from the server to the client to prepopulate the cache. We expect to provide more guidance on this in the future.
